query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+            .field("is_known_empty", &self.is_known_empty)
+            .field("from", &self.from)
+            .field("wheres", &self.wheres)
+            .field("bindings", &self.bindings)
+            .field("known_types", &self.known_types)
+            .field("extracted_types", &self.extracted_types)
+            .finish()
+    }
+}
+
+/// Basics.
+impl ConjoiningClauses {
+    pub fn new() -> ConjoiningClauses {
+        ConjoiningClauses {
+            is_known_empty: false,
+            aliaser: default_table_aliaser(),
+            from: vec![],
+            wheres: vec![],
+            bindings: BTreeMap::new(),
+            known_types: BTreeMap::new(),
+            extracted_types: BTreeMap::new(),
+        }
+    }
+
+    pub fn bind_column_to_var(&mut self, table: TableAlias, column: DatomsColumn, var: Variable) {
+        let alias = QualifiedAlias(table, column);
+
+        // If this is a value, and we don't already know its type or where
+        // to get its type, record that we can get it from this table.
+        let needs_type_extraction =
+            alias.is_value() &&
+            !self.known_types.contains_key(&var) &&
+            !self.extracted_types.contains_key(&var);
+
+        if needs_type_extraction {
+            self.extracted_types.insert(var.clone(), alias.for_type_tag());
+        }
+        self.bindings.entry(var).or_insert(vec![]).push(alias);
+    }
+
+    pub fn constrain_column_to_entity(&mut self, table: TableAlias, column: DatomsColumn, entity: i64) {
+        self.wheres.push(ColumnConstraint::EqualsEntity(QualifiedAlias(table, column), entity))
+    }
+
+    pub fn constrain_attribute(&mut self, table: TableAlias, attribute: i64) {
+        self.constrain_column_to_entity(table, DatomsColumn::Attribute, attribute)
+    }
+
+    pub fn constrain_column_to_constant(&mut self, table: TableAlias, column: DatomsColumn, constant: TypedValue) {
+        self.wheres.push(ColumnConstraint::EqualsValue(QualifiedAlias(table, column), constant))
+    }
+
+    /// Returns false if it's impossible for this type to apply.
+    /// Constraints the var if there's no existing type.
+    fn constrain_var_to_type(&mut self, variable: Variable, this_type: ValueType) -> bool {
+        // Is there an existing binding for this variable?
+        let types_entry = self.known_types.entry(variable);
+        match types_entry {
+            // If so, the types must match.
+            Entry::Occupied(entry) =>
+                *entry.get() == this_type,
+            // If not, record the one we just determined.
+            Entry::Vacant(entry) => {
+                entry.insert(this_type);
+                true
+            }
+        }
+    }
+
+    /// Ensure that the given place has the correct types to be a tx-id.
+    /// Right now this is mostly unimplemented: we fail hard if anything but a placeholder is
+    /// present.
+    fn mark_as_tx(&mut self, tx: &PatternNonValuePlace) {
```

Comment:
Reading these functions, I don't quite see what _mark_ means in this context.  Consider elaborating on "Ensure", here and below.